### ISSUE TITLE
Fall back to canonical constructor in constructor resolution when using Java Records

### ISSUE THREAD

**Issue by vab2048** [spring-projects/spring-data-commons#2625]
Hello,

I have an issue when attempting to persist a record.

The following is fine:

```java
@Table("cat")
public record Cat(
        @Id @Column("id") Long id,
        @Column("name") String name) {

    // Instead of using a constructor uses a factory method.
    public static Cat of(String name) {
        // The 'id' is auto-generated by the DB and populated on insert.
        return new Cat(null, name);
    }

}
```

But when I use another constructor instead of a factory method I get an exception. The code:

```java
@Table("dog")

public record Dog(
        @Id @Column("id") Long id,
        @Column("name") String name) {

    /**
     * Convenience constructor which sets the id to null with the expectation that it will
     * be populated when the entity is inserted into the DB (since it is an auto-incremented column).
     */
    public Dog(String name) {
        this(null, name);
    }
}
```

Test which fails that I would except to pass:

```java
@SpringBootTest
public class DogTest {
    private static final Logger log = LoggerFactory.getLogger(DogTest.class);

    @Autowired
    private JdbcAggregateTemplate jdbcAggregateTemplate;

    // THIS TEST FAILS CURRENTLY.
    @Test
    void animal_insertingIntoDb_succeeds() {
        /* Given: A dog. */
        var fido = new Dog("Fido");

        /* When: We try to insert it in to the DB */
        var insertedFido = jdbcAggregateTemplate.insert(fido);

        /* Then: It should populate the ID field */
        assertThat(insertedFido.id()).isNotNull();
    }
}
```

And the exception:

```
Cannot set property id because no setter, wither or copy constructor exists for class com.example.demo.persistence.Dog!
java.lang.IllegalStateException: Cannot set property id because no setter, wither or copy constructor exists for class com.example.demo.persistence.Dog!
	at org.springframework.data.mapping.model.InstantiationAwarePropertyAccessor.setProperty(InstantiationAwarePropertyAccessor.java:113)
	at org.springframework.data.mapping.model.ConvertingPropertyAccessor.setProperty(ConvertingPropertyAccessor.java:64)
	at org.springframework.data.jdbc.core.JdbcAggregateChangeExecutionContext.setIdAndCascadingProperties(JdbcAggregateChangeExecutionContext.java:337)
	at org.springframework.data.jdbc.core.JdbcAggregateChangeExecutionContext.populateIdsIfNecessary(JdbcAggregateChangeExecutionContext.java:305)
	at org.springframework.data.jdbc.core.AggregateChangeExecutor.execute(AggregateChangeExecutor.java:52)
	at org.springframework.data.jdbc.core.JdbcAggregateTemplate.store(JdbcAggregateTemplate.java:339)
	at org.springframework.data.jdbc.core.JdbcAggregateTemplate.insert(JdbcAggregateTemplate.java:167)
	at com.example.demo.persistence.DogTest.animal_insertingIntoDb_succeeds(DogTest.java:26)
```

Attached is a repo you can run to reproduce the problem: [spring-data-jdbc-records-bug.zip](https://github.com/spring-projects/spring-data-commons/files/8639991/spring-data-jdbc-records-bug.zip)

Am I doing something wrong here (missing annotation/etc.) I would expect the framework to be able to pick up the correct constructor for the record but it currently does not.

**Comment by schauder** [comment]
TL;DR:
Add 
```java
    @PersistenceConstructor
    public Dog {}
```
to your `Dog` record.

Explanation:

When persisting an entity with a database generated ID Spring Data JDBC has to set the id in the entity. It tries to find a setter first. Since a record is immutable it obviously doesn't find any.

It then looks for a "wither" (I'm still looking for a better name for those). Something like `withId(Long id)` returning a new Dog instance. So another fix would be to create one of those.

The final attempt is to use a constructor and us it to construct a copy with the new Id. For `Cat` this works, because there is a single all args constructor. For `Dog` it doesn't because there are two constructors and Spring Data doesn't know which to use.

The proper way to tell Spring Data which constructor to use is with the `@PersistenceConstructor` annotation. With records the next challenge becomes where to put that constructor and the answer is: On a "compact constructor" as demonstrated above. See also: https://stackoverflow.com/a/67168625/66686

**Comment by vab2048** [comment]
> The final attempt is to use a constructor and us it to construct a copy with the new Id. For Cat this works, because there is a single all args constructor. For Dog it doesn't because there are two constructors and Spring Data doesn't know which to use.

I can understand that from the framework perspective. But from a framework user's perspective is it wrong of me to expect the framework to know that since it is a record that if the arity matches then it should, by default, just go for the canonical constructor? (This is because any other constructor, by definition, will just call the canonical constructor anyway (confirmed by the the last part of the spec [here](https://docs.oracle.com/javase/specs/jls/se15/preview/specs/records-jls.html#jls-8.10.4))).

If the default behaviour is undesirable then the framework user can resort to using the `@PersistenceConstructor` annotation. 

Just a suggestion... what do you think?

**Comment by schauder** [comment]
We'll consider that.
